### PR TITLE
ARROW-15410: [C++][Datasets] Improve memory usage of datasets API when scanning parquet

### PR DESCRIPTION
**Pull request by westonpace**
This PR changes a few things.

 * The default file readahead is changed to 4.  This doesn't seem to affect performance on HDD/SSD and users should already be doing special tuning for S3.  Besides, in many cases, users are reading IPC/Parquet files that have many row groups and so we already have sufficient I/O parallelism.  This is important for bringing down the overall memory usage as can be seen in the formula below.
 * The default batch readahead is changed to 16.  Previously, when we were doing filtering and projection within the scanner, it made sense to read many batches ahead (generally want at least 2 * # of CPUs in that case).  Now that the exec plan is doing the computation the exec plan buffering is instead handled by kDefaultBackpressureLowBytes and kDefaultBackpressureHighBytes.
 * Moves around the parquet readahead a bit.  The previous version would read ahead N row groups.  Now we always read ahead exactly 1 row group but we read ahead N batches (this may mean that we read ahead more than 1 row group if the batch size is much larger than the row group size).
 * Backpressure now utilizes the pause/resume producing signals in the execution plan.  I've adding a `counter` argument to the calls to help deal with the challenges that arise when we try and sequence backpressure signals.  Partly this was to add support for monitoring backpressure (for tests).  Partly it is because I have since become more aware of the reasons for these signals.  They are needed to allow for backpressure from the aggregate & join nodes.
 * Sink backpressure can now be monitored.  This makes it easier to test and could be potentially useful to a user that wanted to know when they are consuming the plan too slowly.
 * Changes the default scanner batch size to 128Ki rows.  Now that we have more or less decoupled the scanning batch size from the row group size we can pass smaller batches through the scanner.  This makes it easier to get parallelism on small datasets..

Putting this altogether the scanner should now buffer in memory:

MAX(fragment_readahead * row_group_size_bytes * 2, fragment_readahead * batch_readahead * batch_size_bytes)

The exec plan sink node should buffer ~ kDefaultBackpressureHighBytes bytes.

The exec plan itself can have some number of tasks in flight but, assuming there are no pipeline breakers, this will be limited to the number of threads in the CPU thread pool and so it should be parallelism * batch_size_bytes.

Adding those together should give the total RAM usage of a plan being read via a sink node that doesn't have any pipeline breakers.

When the sink is a write node then there is a separate backpressure consideration based on # of rows (we can someday change this to be # of bytes but it would be a bit tricky at the moment because we need to balance this with the other write parameters like min_rows_per_group).

So, given the parquet dataset mentioned in the JIRA (21 files, 10 million rows each, 10 row groups each) and knowing that 1 row group is ~140MB when decompressed into Arrow format we should get the following default memory usage:

Scanner readahead = MAX(4 * 140MB * 2, 4 * 16 * 17.5MB) = MAX(1120MB, 1120MB) = 1120MB
Sink readahead ~ 1GiB
Total RAM usage should then be ~2GiB.

 - [x] Add tests to verify memory usage
 - [ ] ~~Update docs to mention that S3 users may want to increase the fragment readahead but this will come at the cost of more RAM usage.~~
 - [ ] ~~Update docs to give some of this "expected memory usage" information~~

### CODE REVIEW

**Review by lidavidm — APPROVED** [review]
These generally look like good changes. I wonder how we can test it, perhaps looking at the memory pool?

**Comment by westonpace** [comment]
@lidavidm Unfortunately, when adding tests, I ran into a few issues and so this PR has evolved quite a bit from the last time you reviewed it.  It probably needs a fresh look at this point.

**Comment by westonpace** [comment]
I've created ARROW-16263 for the documentation tasks.

File: cpp/src/arrow/compute/exec/sink_node.cc [hunk 1]
```diff
@@ -46,31 +46,81 @@ using internal::checked_cast;
 namespace compute {
 namespace {
 
+class BackpressureResevoir : public BackpressureMonitor {
```
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/sink_node.cc:49]: nit but seems it's spelled Rese**r**voir

File: cpp/src/arrow/compute/exec/options.h [hunk 2]
```diff
@@ -129,17 +129,86 @@ class ARROW_EXPORT AggregateNodeOptions : public ExecNodeOptions {
...
 class ARROW_EXPORT SinkNodeOptions : public ExecNodeOptions {
  public:
-  explicit SinkNodeOptions(std::function<Future<util::optional<ExecBatch>>()>* generator,
-                           util::BackpressureOptions backpressure = {})
-      : generator(generator), backpressure(std::move(backpressure)) {}
-
+  explicit SinkNodeOptions(
+      std::function<Future<util::optional<ExecBatch>>()>* generator,
+      BackpressureOptions backpressure = {},
+      std::shared_ptr<BackpressureMonitor>* backpressure_monitor = NULLPTR)
```
[the remark this answers is not in the document]
Comment (lidavidm) [reply cpp/src/arrow/compute/exec/options.h:174]: Ah, this is because it's an out pointer?

File: cpp/src/arrow/compute/exec/sink_node.cc [hunk 3]
```diff
@@ -103,6 +157,28 @@ class SinkNode : public ExecNode {
 
   Future<> finished() override { return finished_; }
 
+  void RecordBackpressureBytesUsed(const ExecBatch& batch) {
+    if (backpressure_queue_->enabled()) {
+      uint64_t bytes_used = static_cast<uint64_t>(batch.TotalBufferSize());
+      auto state_change = backpressure_queue_->RecordProduced(bytes_used);
+      if (state_change >= 0) {
+        EVENT(span_, "Backpressure applied", {{"counter", state_change}});
```
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/sink_node.cc:165]: nit, but maybe namespace the span attribute name?

File: cpp/src/arrow/compute/exec/sink_node.cc [hunk 4]
```diff
@@ -46,31 +46,81 @@ using internal::checked_cast;
...
                  /*num_outputs=*/0),
-        producer_(MakeProducer(generator, std::move(backpressure))) {}
+        backpressure_queue_(std::make_shared<BackpressureResevoir>(
+            backpressure.resume_if_below, backpressure.pause_if_above)),
+        push_gen_(),
+        producer_(push_gen_.producer()) {
+    if (backpressure_monitor_out) {
+      *backpressure_monitor_out = backpressure_queue_;
+    }
+    AsyncGenerator<util::optional<ExecBatch>> captured_gen = push_gen_;
```
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/sink_node.cc:104]: nit, but if we're capturing `this` already, is it necessary to also copy/capture the generator separately?

File: cpp/src/arrow/compute/exec/sink_node.cc [hunk 5]
```diff
@@ -46,31 +46,81 @@ using internal::checked_cast;
...
+      : bytes_used_(0),
+        state_change_counter_(0),
+        resume_if_below_(resume_if_below),
+        pause_if_above_(pause_if_above) {}
+
+  uint64_t bytes_in_use() const override { return bytes_used_; }
+  bool is_paused() const override { return state_change_counter_ % 2 == 1; }
+  bool enabled() const { return pause_if_above_ > 0; }
+
+  int32_t RecordProduced(uint64_t num_bytes) {
```
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/sink_node.cc:61]: It seems the exact return value is never used, perhaps return bool instead?

**Review by lidavidm — APPROVED** [review]
Thanks for working this out. I left some small comments but overall this looks pretty straightforward and doesn't involve complex generators which is a plus.

File: cpp/src/arrow/compute/exec/hash_join_node.cc [hunk 6]
```diff
@@ -593,9 +593,13 @@ class HashJoinNode : public ExecNode {
     return Status::OK();
   }
 
-  void PauseProducing(ExecNode* output) override { EVENT(span_, "PauseProducing"); }
+  void PauseProducing(ExecNode* output, int32_t counter) override {
+    // TODO(ARROW-16246)
```
[the remark this answers is not in the document]
Comment (westonpace) [reply cpp/src/arrow/compute/exec/hash_join_node.cc:597]: I'm not really a fan of these if we're not actually doing anything (or just proxying info) but I can put it back in.  In general I think `EVENT` is more meaningful when we've made some kind of decision based on data (e.g. when we decide we need to apply backpressure) or an external signal.

File: cpp/src/arrow/compute/exec/options.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (westonpace) [reply cpp/src/arrow/compute/exec/options.h:174]: It's an out pointer that we want to share ownership.  I was a little torn on this as we could use a raw pointer but then it wouldn't be valid to call any of the monitor methods after the plan is destroyed.  I think that is probably ok in general but `generator` _will_ actually keep the queue alive beyond the plan (`std::function` is essentially a shared pointer to its target) and so I wanted to match the lifetime.

File: cpp/src/arrow/compute/exec/sink_node.cc (hunk 5 above)
Comment (westonpace) [reply cpp/src/arrow/compute/exec/sink_node.cc:61]: Which return value are you referring to?  The return value from `RecordProduced` is important as its the signal we send through the plan:

```
      auto state_change = backpressure_queue_.RecordProduced(bytes_used);
      if (state_change >= 0) {
        EVENT(span_, "Backpressure applied", {{"backpressure.counter", state_change}});
        inputs_[0]->PauseProducing(this, state_change);
      }
```

...and then later in the source node it shows up as `counter`...

```
  void PauseProducing(ExecNode* output, int32_t counter) override {
    std::lock_guard<std::mutex> lg(mutex_);
    if (counter <= backpressure_counter_) {
      return;
    }
```

File: cpp/src/arrow/compute/exec/options.h [hunk 7]
```diff
@@ -129,17 +129,86 @@ class ARROW_EXPORT AggregateNodeOptions : public ExecNodeOptions {
   std::vector<FieldRef> keys;
 };
 
+constexpr int32_t kDefaultBackpressureHighBytes = 1 << 30;  // 1GiB
+constexpr int32_t kDefaultBackpressureLowBytes = 1 << 28;   // 256MiB
+
+class BackpressureMonitor {
```
[the remark this answers is not in the document]
Comment (westonpace) [reply cpp/src/arrow/compute/exec/options.h]: Thanks.  Added.

File: cpp/src/arrow/compute/exec/options.h [hunk 8]
```diff
@@ -129,17 +129,86 @@ class ARROW_EXPORT AggregateNodeOptions : public ExecNodeOptions {
...
+  /// and backpressure can be applied even if this is not used.
+  std::shared_ptr<BackpressureMonitor>* backpressure_monitor;
+};
+
+/// \brief Control used by a SinkNodeConsumer to pause & resume
+///
+/// Callers should ensure that they do not call Pause and Resume simultaneously and they
+/// should sequence things so that a call to Pause() is always followed by an eventual
+/// call to Resume()
+class BackpressureControl {
```
[the remark this answers is not in the document]
Comment (westonpace) [reply cpp/src/arrow/compute/exec/options.h]: Added.

File: cpp/src/arrow/compute/exec/options.h [hunk 9]
```diff
@@ -129,17 +129,86 @@ class ARROW_EXPORT AggregateNodeOptions : public ExecNodeOptions {
...
+  ///                       queue has more than pause_if_above items
+  BackpressureOptions(uint32_t resume_if_below, uint32_t pause_if_above)
+      : resume_if_below(resume_if_below), pause_if_above(pause_if_above) {}
+
+  static BackpressureOptions DefaultBackpressure() {
+    return BackpressureOptions(kDefaultBackpressureLowBytes,
+                               kDefaultBackpressureHighBytes);
+  }
+
+  inline bool should_apply_backpressure() const { return pause_if_above > 0; }
```
[the remark this answers is not in the document]
Comment (westonpace) [reply cpp/src/arrow/compute/exec/options.h]: Nope, removed.

File: cpp/src/arrow/compute/exec/sink_node.cc (hunk 1 above)
Comment (westonpace) [reply cpp/src/arrow/compute/exec/sink_node.cc]: :exploding_head: Fixed...but this is a serious [Berenstain Bears](https://en.wikipedia.org/wiki/Berenstain_Bears#Name_confusion) moment.

File: cpp/src/arrow/compute/exec/sink_node.cc (hunk 3 above)
Comment (westonpace) [reply cpp/src/arrow/compute/exec/sink_node.cc]: I changed it to `backpressure.counter`.  Is this what you were thinking?  It doesn't seem we are doing much namespacing elsewhere.

File: cpp/src/arrow/compute/exec/sink_node.cc (hunk 4 above)
Comment (westonpace) [reply cpp/src/arrow/compute/exec/sink_node.cc]: No, I've changed this.

File: cpp/src/arrow/compute/exec/sink_node.cc (hunk 5 above)
Comment (lidavidm) [reply cpp/src/arrow/compute/exec/sink_node.cc:61]: Ah, my bad, I missed that spot.

File: cpp/src/arrow/compute/exec/sink_node.cc (hunk 3 above)
Comment (lidavidm) [reply cpp/src/arrow/compute/exec/sink_node.cc]: That works, I was thinking `arrow....` like in the memory PR but it's not a big deal.

File: cpp/src/arrow/compute/exec/hash_join_node.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (lidavidm) [reply cpp/src/arrow/compute/exec/hash_join_node.cc:597]: Ah, that's fair. We can leave it out.

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
Looks like Raúl is still working through the Git failures, but we can rebase and merge after that.

**Comment by westonpace** [comment]
@lidavidm 

Hmm....doing some more testing on this I think this might be non-ideal in a few situations (S3, low number of files, smallish row groups).  This is because we are always reading only 1 row group ahead so we will read, at most, 2 reads in parallel for a single file.

However, the old behavior was also unmaintainable as it would have kicked off dozens of parallel reads and run out of memory.

The ideal approach would be to keep track of how many rows we have "in flight" and issue reads until we have batch_size * batch_readahead rows in flight and then pause.  I'm going to work on this but, as we are close to release, I'd prefer to move forward with this sometimes-slower-but-usually-safer approach and put the ideal fix in a follow-up.

**Comment by lidavidm** [comment]
Is there a quick config change we can document for S3?

**Comment by lidavidm** [comment]
Just so that regressions can be easily fixed

**Comment by lidavidm** [comment]
That said you should feel free to merge here in that case, we should prefer not crashing/freezing

**Comment by westonpace** [comment]
There's no config that we can easily change that would go back to the old behavior.

**Comment by lidavidm** [comment]
Hmm, okay.

Well, low number of files + smallish row groups should hopefully not be too big a deal then. It might be worth calling this out in the release notes come time just so that it's clear we're working on this.

**Comment by westonpace** [comment]
Ok.  I'm going to merge this so we get some nightly tests, etc. run on it and then I'm going to see if I can get a PR up for the ideal behavior pretty quickly.

**Comment by kou** [comment]
@westonpace This breaks a GLib CI on macOS:

https://github.com/apache/arrow/runs/6136118707?check_suite_focus=true#step:9:128

```text
FAILED: arrow-glib/libarrow-glib.800.dylib.p/expression.cpp.o 
ccache c++ -Iarrow-glib/libarrow-glib.800.dylib.p -Iarrow-glib -I../../c_glib/arrow-glib -I. -I../../c_glib -I/usr/local/include -I/usr/local/Cellar/thrift/0.16.0/include -I/usr/local/Cellar/brotli/1.0.9/include -I/usr/local/Cellar/protobuf/3.19.4/include -I/usr/local/Cellar/lz4/1.9.3/include -I/usr/local/Cellar/zstd/1.5.2/include -I/usr/local/Cellar/utf8proc/2.7.0/include -I/usr/local/Cellar/libffi/3.4.2/include -I/usr/local/Cellar/glib/2.72.1/include -I/usr/local/Cellar/glib/2.72.1/include/glib-2.0 -I/usr/local/Cellar/glib/2.72.1/lib/glib-2.0/include -I/usr/local/opt/gettext/include -I/usr/local/Cellar/pcre/8.45/include -fcolor-diagnostics -Wall -Winvalid-pch -Wnon-virtual-dtor -Werror -std=c++11 -O0 -g -Wmissing-declarations -DARROW_NO_DEPRECATED_API -DUTF8PROC_EXPORTS -MD -MQ arrow-glib/libarrow-glib.800.dylib.p/expression.cpp.o -MF arrow-glib/libarrow-glib.800.dylib.p/expression.cpp.o.d -o arrow-glib/libarrow-glib.800.dylib.p/expression.cpp.o -c ../../c_glib/arrow-glib/expression.cpp
In file included from ../../c_glib/arrow-glib/expression.cpp:20:
In file included from ../../c_glib/arrow-glib/compute.hpp:22:
In file included from /usr/local/include/arrow/compute/api.h:48:
/usr/local/include/arrow/compute/exec/options.h:202:20: error: 'arrow::compute::BackpressureControl' has virtual functions but non-virtual destructor [-Werror,-Wnon-virtual-dtor]
class ARROW_EXPORT BackpressureControl {
                   ^
1 error generated.
```

Could you confirm this? (Is `virtual ~BackpressureControl() = default;` missing?)

**Comment by lidavidm** [comment]
Sorry, we should've looked at CI more carefully. Thanks for catching this @kou and please see #12968

**Comment by westonpace** [comment]
Sorry Kou.  Thanks David!

**Pull request closed without merging**